_posts/2025-04-11-Hybrid-Search-Options-And-Practice.md

```diff
@@ -0,0 +1,275 @@
+  "phase_results_processors": [
+    {
+      "normalization-processor": {
+        "normalization": {
+          "technique": "min_max"
+        },
+        "combination": {
+          "technique": "arithmetic_mean",
+          "parameters": {
+            "weights": [
+              0.3,
+              0.7
+            ]
+          }
+        }
+      }
+    }
+  ]
+}
+```
+
+Use the same search pipeline with the search request
+
+```
+GET /bookstore_catalog/_search?search_pipeline=min_max-search-pipeline
+{
+  "_source": {
+    "exclude": [
+      "title_embedding"
+    ]
+  },
+  "query": {
+    "hybrid": {
+      "queries": [
+        {
+          "match": {
+            "title": {
+              "query": "science fiction"
+            }
+          }
+        },
+        {
+          "neural": {
+            "title_embedding": {
+              "query_text": "science fiction",
+              "model_id": "aVeif4oB5Vm0Tdw8zYO2",
+              "k": 5
+            }
+          }
+        }
+      ]
+    }
+  }
+}
+
+
+```
+
+
+This query uses:
+
+* `hybrid`: OpenSearch's built-in hybrid query type
+* `search_pipeline`: Applies the min-max normalization and score combination
+
+
+How it works:
+
+1. OpenSearch executes both the `match` and `neural` queries.
+2. The `min_max-search-pipeline` normalizes scores using min-max scaling.
+3. Normalized scores are combined using a weighted arithmetic mean (30% lexical, 70% neural in this case which were defined in the search pipeline).
+
+
+**When to use score normalization:**
+
+* **When You Can Calibrate or Train Weights:** If you have a reliable way to normalize scores (e.g. using min-max or L2 norms) and **tune the weight of each retrieval method**, a linear score combination can slightly outperform rank fusion in relevance metrics
+* **Emphasizing a Superior Signal:** Score normalization allows you to explicitly emphasize the more reliable model for your use case. In a general document search, some queries might be answered mostly by keyword matching while others need semantic understanding. By adjusting weights (or using query-dependent logic), you can give BM25 or the neural model more influence when appropriate. 
+* **Controlled Merging (Fewer False Positives):** A normalized linear combination gives more continuous control over how results merge, which can reduce the chance of a less relevant document being boosted solely due to one method. For instance, if a semantic model returns a somewhat off-target result at a high rank, RRF would still elevate it because of its rank. A score-based approach can mitigate this by yielding a lower normalized relevance score for that document (since BM25 might give it 0, and the semantic score alone may not be enough to beat other combined scores). Thus, in a well-tuned hybrid system, score normalization can lead to a more precise top-10 ranking, as each document’s final score reflects a balance of both lexical and semantic relevance.
```

Comment:
what does this mean: "gives more continuous control over how results merge"? specifically I'm not sure about the word `continuous`, I think we can drop it from this sentence